### PR TITLE
NameCheap DNS provider

### LINKED ISSUE #376: NameCheap DNS validation
Hi there. I'm currently using a custom script to handle this, but it seems like it would be relatively easy to implement DNS validation for NameCheap. Their API is pretty straightforward.

https://www.namecheap.com/support/api/methods/domains-dns.aspx

### PR DESCRIPTION
**Pull request by impworks**
Closes #376.

Tricky things to know about NameCheap API:

* API is only available if you have $50 on your account or have bought $50 worth of services in the last 2 years
* The "API activation takes up to 2 business days" claim is baloney, they can do it immediately if you contact support and will likely ignore your request altogether if you don't.
* You need to manually add your server's IP to a whitelist before you can send API requests. If you have a dynamic IP, well, oops, you're out of luck.
* The API _also_ requires sending current client IP with every single request (hence the "IP" field in NameCheap credentials).

I didn't add anything to the documentation project yet. Can open another pull request there if required.

### CODE REVIEW

**Comment by webprofusion-chrisc** [comment]
Thanks! I have added a few questions/comments. In particular we need to be doubly sure we only remove TXT records, and we also need to make sure we can handle 3 level domains like domain.co.uk, *.domain.co.uk (vs domain.com)

**Comment by impworks** [comment]
@webprofusion-chrisc hmm, I don't see any review comments in the code

File: src/Certify.Providers/DNS/NameCheap/DnsProviderNameCheap.cs [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        /// <summary>
+        /// Removes a DNS record from the list.
+        /// </summary>
+        public async Task<ActionResult> DeleteRecord(DnsRecord request)
+        {
+            try
+            {
+                var hosts = await GetHosts(request.TargetDomainName);
+                var recordName = request.RecordName.Replace(ParseDomain(request.TargetDomainName).Mask, "");
+                var toRemove = hosts.FirstOrDefault(x => x.Name == recordName && x.Address == request.RecordValue);
```
[the remark this answers is not in the document]
Comment (impworks) [reply src/Certify.Providers/DNS/NameCheap/DnsProviderNameCheap.cs]: Right, added an extra check for `RecordType`.

File: src/Certify.Providers/DNS/NameCheap/DnsProviderNameCheap.cs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        /// <summary>
+        /// Returns the list of hosts for a domain.
+        /// </summary>
+        private async Task<List<NameCheapHostRecord>> GetHosts(string domain)
+        {
+            var domainInfo = ParseDomain(domain);
+
+            var xml = await InvokeGetApiAsync("domains.dns.getHosts", new Dictionary<string, string>
+            {
+                ["SLD"] = domainInfo.SLD,
```
[the remark this answers is not in the document]
Comment (impworks) [reply src/Certify.Providers/DNS/NameCheap/DnsProviderNameCheap.cs]: Good question, I'll have to ask their help support

**Comment by impworks** [comment]
@webprofusion-chrisc I updated the PR, fixing both problems you pointed out.

**Comment by webprofusion-chrisc** [comment]
Awesome! Thanks for all your hard work :)

**Comment by impworks** [comment]
@webprofusion-chrisc No problems, I needed it anyway, and I hope some other NameCheap client will find it useful as well :)

**Pull request merged**